File: Assignment_Spring_Animation/app/src/main/java/com/example/springanimation/MainActivity.java
package com.example.springanimation;

import android.os.Bundle;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.dynamicanimation.animation.SpringAnimation;

public class MainActivity extends AppCompatActivity {

    SpringAnimation springAnimation;
    float anim = 100;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ImageView image = findViewById(R.id.springer);

        springAnimation = new SpringAnimation(image,SpringAnimation.SCROLL_Y,anim);

        image.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                Log.d("beep","beep");
                springAnimation.start();
                anim = -anim;
                springAnimation.animateToFinalPosition(anim);
                return false;
            }
        });
    }
}

File: WeatherApplication/app/src/main/java/com/example/weatherapplication/view/MainActivity.kt
package com.example.weatherapplication.view

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.weatherapplication.R
import com.example.weatherapplication.model.WeatherAPI
import com.example.weatherapplication.model.entities.Weather
import com.example.weatherapplication.model.entities.WeatherResponse
import com.example.weatherapplication.viewmodel.WeatherViewModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.launch
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory


class MainActivity : AppCompatActivity() {

    private lateinit var adapter: WeatherAdapter
    private lateinit var recyclerView: RecyclerView
    private lateinit var viewManager: RecyclerView.LayoutManager

    private lateinit var weatherViewModel: WeatherViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        Log.d("beep", "Got to main activity")

        viewManager = LinearLayoutManager(this)

        adapter = WeatherAdapter(emptyList())

        recyclerView = findViewById<RecyclerView>(R.id.weather_list).apply{
            // use this setting to improve performance if you know that changes
            // in content do not change the layout size of the RecyclerView
            setHasFixedSize(true)
            layoutManager = viewManager
            this.adapter = this@MainActivity.adapter
        }
        weatherViewModel = ViewModelProvider(this).get(WeatherViewModel::class.java)
        weatherViewModel.weather.observe(this, Observer {weather ->
            weather?.let {
                adapter.setWeather(it)
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        val inflater: MenuInflater = menuInflater
        inflater.inflate(R.menu.weather_menu, menu)
        return true
    }
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle item selection
        return when (item.itemId) {
            R.id.weather_refresh -> {
                weatherViewModel.refreshWeatherMenuItem()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
